calyx-opt/src/passes/compile_invoke.rs

```diff
@@ -131,38 +132,59 @@ impl CompileInvoke {
     ///
     /// Since this pass eliminates all ref cells in post order, we expect that
     /// invoked component already had all of its ref cells removed.
+
     fn ref_cells_to_ports<T>(
         &mut self,
         inv_cell: RRC<ir::Cell>,
         ref_cells: impl Iterator<Item = (ir::Id, ir::RRC<ir::Cell>)>,
     ) -> Vec<ir::Assignment<T>> {
         let inv_comp = inv_cell.borrow().type_name().unwrap();
         let mut assigns = Vec::new();
-        for (ref_cell_name, cell) in ref_cells {
+        for (ref_cell_name, in_cell) in ref_cells {
             log::debug!(
                 "Removing ref cell `{}` with {} ports",
                 ref_cell_name,
-                cell.borrow().ports.len()
+                in_cell.borrow().ports.len()
             );
 
             // Mapping from canonical names of the ports of the ref cell to the
-            // new port defined on the signature of the component
+            // new port defined on the signature of the component. This has name of ref cell, not arg cell
             let Some(comp_ports) = self.port_names.get(&inv_comp) else {
                 unreachable!("component `{}` invoked but not already visited by the pass", inv_comp)
             };
 
-            // The type of the cell is the same as the ref cell so we can
-            // iterate over its ports and generate bindings for the ref cell.
-            for pr in &cell.borrow().ports {
+            // We expect each canonical port in `comp_ports` to exactly match with a port in
+            //`in_cell` based on well-formedness subtype checks.
+            for canon in comp_ports.keys() {
+                //only interested in ports attached to the ref cell
+                if canon.cell != ref_cell_name {
+                    continue;
+                }
+                let in_cell_borrowed = in_cell.borrow();
```

Comment:
It's a common pattern in Rust to use shadowing for situations like this, where you "upgrade" the type of something but it still sorta means the same thing. As in:

```rs
let in_cell = in_cell.borrow();
```